lib/Conversion/ExportVerilog/ExportVerilog.cpp

```diff
@@ -2969,6 +2980,292 @@ SubExprInfo ExprEmitter::visitUnhandledExpr(Operation *op) {
+    addSeparator = true;
+    emitNestedProperty(input, PropertyPrecedence::And);
+  }
+  return {PropertyPrecedence::And};
+}
+
+EmittedProperty PropertyEmitter::visitLTL(ltl::OrOp op) {
+  bool addSeparator = false;
+  for (auto input : op.getInputs()) {
+    if (addSeparator)
+      ps << PP::space << "or" << PP::space;
+    addSeparator = true;
+    emitNestedProperty(input, PropertyPrecedence::Or);
+  }
+  return {PropertyPrecedence::Or};
+}
+
+EmittedProperty PropertyEmitter::visitLTL(ltl::DelayOp op) {
+  ps << "##";
+  if (auto length = op.getLength()) {
+    if (*length == 0) {
+      ps.addAsString(op.getDelay());
+    } else {
+      ps << "[";
+      ps.addAsString(op.getDelay());
+      ps << ":";
+      ps.addAsString(op.getDelay() + *length);
+      ps << "]";
+    }
+  } else {
+    if (op.getDelay() == 0) {
+      ps << "[*]";
+    } else if (op.getDelay() == 1) {
+      ps << "[+]";
+    } else {
+      ps << "[";
+      ps.addAsString(op.getDelay());
+      ps << ":$]";
+    }
+  }
+  ps << PP::space;
+  emitNestedProperty(op.getInput(), PropertyPrecedence::Concat);
+  return {PropertyPrecedence::Concat};
+}
+
+void PropertyEmitter::emitLTLConcat(ValueRange inputs) {
+  bool addSeparator = false;
+  for (auto input : inputs) {
+    if (addSeparator) {
+      ps << PP::space;
+      if (!input.getDefiningOp<ltl::DelayOp>())
+        ps << "##0" << PP::space;
+    }
+    addSeparator = true;
+    emitNestedProperty(input, PropertyPrecedence::Concat);
+  }
+}
+
+EmittedProperty PropertyEmitter::visitLTL(ltl::ConcatOp op) {
+  emitLTLConcat(op.getInputs());
+  return {PropertyPrecedence::Concat};
+}
+
+EmittedProperty PropertyEmitter::visitLTL(ltl::NotOp op) {
+  ps << "not" << PP::space;
+  emitNestedProperty(op.getInput(), PropertyPrecedence::Unary);
+  return {PropertyPrecedence::Unary};
+}
+
+/// For a value `concat(..., delay(const(true), 1, 0))`, return `...`. This is
+/// useful for emitting `(seq ##1 true) |-> prop` as `seq |=> prop`.
+static ValueRange getNonOverlappingConcatSubrange(Value value) {
+  auto concatOp = value.getDefiningOp<ltl::ConcatOp>();
+  if (!concatOp || concatOp.getInputs().size() < 2)
+    return {};
+  llvm::errs() << "got concat " << concatOp << "\n";
+  auto delayOp = concatOp.getInputs().back().getDefiningOp<ltl::DelayOp>();
+  if (!delayOp || delayOp.getDelay() != 1 || delayOp.getLength() != 0)
+    return {};
+  llvm::errs() << "got delay " << delayOp << "\n";
+  auto constOp = delayOp.getInput().getDefiningOp<ConstantOp>();
+  if (!constOp || !constOp.getValue().isOne())
+    return {};
+  llvm::errs() << "got const " << constOp << "\n";
+  return concatOp.getInputs().drop_back();
+}
+
+EmittedProperty PropertyEmitter::visitLTL(ltl::ImplicationOp op) {
+  // Emit `(seq ##1 true) |-> prop` as `seq |=> prop`.
+  if (auto range = getNonOverlappingConcatSubrange(op.getAntecedent());
+      !range.empty()) {
+    emitLTLConcat(range);
+    ps << PP::space << "|=>" << PP::space;
```

Comment:
That `emitAssignLike` is really useful. I wonder if we should move some of these nice helpers into a wrapper around `ps`, such that it's not tied to `{Stmt,Expr,Property}Emitter`, but available anywhere where `ps` are sold.

Reply:
I just realized that we might want to use the pattern of `emitBinary` here, since `|->` and `|=>` are just binary operators, not really assignment-like.

Reply:
Oh, okay that makes sense.  And yes 100% should move those helpers (and maybe some new ones) for reuse and hopefully to make it easier to produce good output / harder to produce bad output :+1: .